File: react-native/android/src/main/java/com/digia/engage/rn/RNEventBridgePlugin.kt
/**
 * RNEventBridgePlugin
 *
 * The single native DigiaCEPPlugin used in React Native.
 *
 * When Digia.register(rnPlugin) is called, the SDK calls setup(delegate) which gives us the
 * DigiaCEPDelegate reference. This delegate is used by triggerCampaign / invalidateCampaign bridge
 * methods to push campaigns into the native rendering engine.
 *
 * Overlay lifecycle events (impressed / clicked / dismissed) received via notifyEvent() are emitted
 * to JS as DeviceEventEmitter events so that JS CEP plugins can report analytics back to their
 * platform.
 */
package com.digia.engage.rn

import com.digia.engage.DiagnosticReport
import com.digia.engage.DigiaCEPDelegate
import com.digia.engage.DigiaCEPPlugin
import com.digia.engage.DigiaExperienceEvent
import com.digia.engage.InAppPayload
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.modules.core.DeviceEventManagerModule

internal class RNEventBridgePlugin(
        private val reactContext: ReactApplicationContext,
) : DigiaCEPPlugin {

    override val identifier: String = "rn-event-bridge"

    /** Delegate received from the SDK via setup(). Used by DigiaModule bridge methods. */
    var delegate: DigiaCEPDelegate? = null
        private set

    private fun emit(event: String, params: com.facebook.react.bridge.WritableMap) {
        if (reactContext.hasActiveReactInstance()) {
            reactContext
                    .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
                    .emit(event, params)
        }
    }

    override fun setup(delegate: DigiaCEPDelegate) {
        this.delegate = delegate
    }

    override fun forwardScreen(name: String) {
        /* forwarded by Digia.setCurrentScreen() on the native side */
    }

    override fun notifyEvent(event: DigiaExperienceEvent, payload: InAppPayload) {
        val params =
                Arguments.createMap().apply {
                    putString("campaignId", payload.id)
                    when (event) {
                        is DigiaExperienceEvent.Impressed -> putString("type", "impressed")
                        is DigiaExperienceEvent.Clicked -> {
                            putString("type", "clicked")
                            event.elementId?.let { putString("elementId", it) }
                        }
                        is DigiaExperienceEvent.Dismissed -> putString("type", "dismissed")
                    }
                }
        emit("digiaOverlayEvent", params)
    }

    override fun teardown() {
        delegate = null
    }

    override fun healthCheck(): DiagnosticReport =
            DiagnosticReport(isHealthy = true, metadata = mapOf("identifier" to identifier))
}

File: react-native/ios/RNEventBridgePlugin.swift
/**
 * RNEventBridgePlugin
 *
 * Implements DigiaCEPPlugin so it can be registered with the native iOS Digia
 * SDK via Digia.register(). It acts as the bridge between the native SDK's
 * DigiaCEPDelegate and the JS layer.
 *
 * Lifecycle
 * ─────────
 * 1. JS calls Digia.register() → native registerBridge() → this plugin is
 *    registered with the native SDK via Digia.register(rnPlugin).
 * 2. The SDK calls setup(delegate:) storing the delegate reference.
 * 3. When a JS CEP plugin delivers a campaign via onCampaignTriggered(),
 *    the JS side calls triggerCampaign() which forwards here via DigiaModule.
 * 4. Overlay lifecycle events (impressed / clicked / dismissed) travel back to
 *    JS as DeviceEventEmitter events.
 */
import Foundation
import React
import DigiaEngage

@objc
internal final class RNEventBridgePlugin: NSObject, DigiaCEPPlugin {

    let identifier = "com.digia.rn.bridge"

    /// Populated by the SDK when registered. Used to forward JS-side campaigns
    /// into the native overlay rendering pipeline.
    private(set) weak var delegate: DigiaCEPDelegate?

    private weak var eventEmitter: RCTEventEmitter?

    init(eventEmitter: RCTEventEmitter?) {
        self.eventEmitter = eventEmitter
    }

    // MARK: - DigiaCEPPlugin

    func setup(delegate: DigiaCEPDelegate) {
        self.delegate = delegate
    }

    func notifyEvent(_ event: DigiaExperienceEvent, payload: CEPTriggerPayload) {

        var body: [String: Any] = ["campaignId": payload.cepCampaignId]
        switch event {
        case .impressed:
            body["type"] = "impressed"
        case .clicked(let elementID):
            body["type"] = "clicked"
            if let elementID {
                body["elementId"] = elementID
            }
        case .dismissed:
            body["type"] = "dismissed"
        @unknown default:
            // Unknown future event kind — nothing meaningful to forward.
            return
        }
        let hasEmitter = eventEmitter != nil
        print("[DigiaRN] RNEventBridgePlugin.notifyEvent type=\(body["type"] ?? "?") campaignId=\(payload.cepCampaignId) hasEmitter=\(hasEmitter)")
        eventEmitter?.sendEvent(withName: "digiaEngageEvent", body: body)
    }

    /// Forwards the action to the JS layer (the host handles routing), so it
    /// reports the action as handled — the native open fallback is skipped.
    func notifyAction(actionType: String, url: String, payload: CEPTriggerPayload) -> Bool {
        let body: [String: Any] = [
            "campaignId": payload.cepCampaignId,
            "type": "action",
            "actionType": actionType,
            "url": url,
        ]
        eventEmitter?.sendEvent(withName: "digiaEngageEvent", body: body)
        return true
    }

    func healthCheck() -> DiagnosticReport {
        return DiagnosticReport(
            isHealthy: delegate != nil,
            issue: delegate == nil ? "No delegate set" : nil,
            resolution: delegate == nil ? "Call registerBridge() before triggerCampaign()" : nil
        )
    }

    func teardown() {
        delegate = nil
    }
}
